tensortrade/stochastic/processes/gan.py

```diff
@@ -0,0 +1,381 @@
+
+        return df
+
+    def prices_real_data_frame(self):
+        df = pd.DataFrame()
+
+        for i in range(self._n_samples):
+            df[i] = self.price_real()
+
+        return df
+
+    def train_generator(self, batch_idx, offset):
+        # update the generator
+        for g in range(self._g_rounds):
+            Y_mb = self.get_batch(self._samples, self._batch_size, batch_idx + g + offset)
+            _, G_loss_curr = self._sess.run([self._G_solver, self._G_loss],
+                                            feed_dict={self._CG: Y_mb,
+                                                       self._Z: self.sample_Z(self._batch_size, self._seq_length,
+                                                                              self._latent_dim)})
+        return G_loss_curr
+
+    def train_discriminator(self, batch_idx, offset):
+        # update the discriminator
+        for d in range(self._d_rounds):
+            # using same input sequence for both the synthetic data and the real one,
+            # probably it is not a good idea...
+            X_mb = self.get_batch(self._samples, self._batch_size, batch_idx + d + offset)
+            X_mb = X_mb.reshape(self._batch_size, self._seq_length, 1)
+            Y_mb = self.get_batch(self._samples, self._batch_size, batch_idx + d + offset)
+            _, D_loss_curr = self._sess.run([self._D_solver, self._D_loss],
+                                            feed_dict={self._CD: Y_mb, self._CG: Y_mb, self._X: X_mb,
+                                                       self._Z: self.sample_Z(self._batch_size, self._seq_length,
+                                                                              self._latent_dim)})
+        return D_loss_curr
+
+    def _initialize_gan(self):
+        tf.reset_default_graph()
+
+        self._cal_cagr_n_vol()
+
+        self._samples = self.sample_data()
+        self._initialize_placeholders()
+
+        # Define the loss function for the GAN
+        self._g_sample = self.generator(self._Z, self._CG)
+        d_real, d_logit_real = self.discriminator(self._X, self._CD)
+        d_fake, d_logit_fake = self.discriminator(self._g_sample, self._CG, reuse=True)
+
+        generator_vars = [v for v in tf.trainable_variables() if v.name.startswith('generator')]
+        discriminator_vars = [v for v in tf.trainable_variables() if v.name.startswith('discriminator')]
+
+        D_loss_real = tf.reduce_mean(tf.nn.sigmoid_cross_entropy_with_logits(logits=d_logit_real,
+                                                                             labels=tf.ones_like(d_logit_real)))
+        D_loss_fake = tf.reduce_mean(tf.nn.sigmoid_cross_entropy_with_logits(logits=d_logit_fake,
+                                                                             labels=tf.zeros_like(d_logit_fake)))
+        self._D_loss = D_loss_real + D_loss_fake
+        self._G_loss = tf.reduce_mean(tf.nn.sigmoid_cross_entropy_with_logits(logits=d_logit_fake,
```

Comment:
Also we should allow for loss to be defined by the user if they so choose. For example, the Wasserstein metric leads to better convergence in GANs and would be a good choice for stabilizing the training process so I should be able to give that as a loss.